Feature/advanced search query formulation support

aristotelisxs opened the pull request:
A first version of the advanced search endpoint, separate to the core_search endpoint.

### Description

We separated the two endpoints (advanced and simple search) here for consistency/maintainability and also because it was more convenient to work with data.frame types for advanced search queries instead of character strings. The parsing of advanced queries checks each "acceptable" filter included in the query and performs conversions to match Elasticsearch query syntax for either including, excluding string terms, performing field specific searches, etc.

### Related Issue

Fix for #4.

### Example

The following will accept a batch request that will perform a search for "data mining" and "machine learning", excluding the terms "social", "science" and "medicine" respectively. Year ranges are also specified as 2013 to 2014 and 2000 to 2016 respectively.

```
query <- data.frame("all_of_the_words" = c("data mining", "machine learning"),
"without_the_words" = c("social science", "medicine"),"year_from" = c("2013","2000"),  
"year_to" = c("2014", "2016"))
```

res <- core_advanced_search(query)
head(res$data)
res$data[[1]]$id

aristotelisxs committed:
First version of advanced query formulation functions after direct translation from Typescript on CORE’s backend. Still unoptimised, further commits to follow on this.

aristotelisxs committed:
paste3 would fall apart if there were 3 or more consecutive NA variables and would add an extra entry with an “AND” separator which, in return, would break CORE’s search engine. The replacement function will omit was, given na.rm = T (default) and can be extended to dropping empty strings by replacing na.omit(x) with x[!is.na(x) & !(x %in% "")] if desired.

aristotelisxs committed:
Adding redirection control for advanced and simple queries in the “core_search” function. However, the query needs/is preferable in data frame format instead of character (array), so it would be confusing to leave it as that here.
Also, a small correction on the acceptable advanced query filters.

Next commit to create another endpoint for advanced searches for consistency over data types.

aristotelisxs committed:
Separating documentation and implementation for advanced_search from the search endpoint because of the aforementioned differences in required different types and also for the sake of maintainability/consistency.

aristotelisxs committed:
1) Move operations on advanced_query to zzz.R instead of the advanced search endpoint
2) Add examples for viewing results from advanced search (also make variable names shorter)

sckott commented:
thanks!

the link https://core.ac.uk/docs/#!/all/advanced_search doesn't seem to link to anything for me, where should that link to?

sckott commented:
- tidy line widths to 80 chars or less 
- include a check that the input is a data.frame, since `apply()` i don't think fails well if e.g. a list is given

aristotelisxs committed:
1) Includes details on the “query” dataframe structure to be supplied for advanced search queries (as there is an absence of CORE related pages to explain this)
2) Fixing line limits to 80 characters each
3) Minor changes to the core_search include removing a reference to the non-existent searchBatch page and double to single quote strings in the examples section (for styling consistency)

aristotelisxs committed:
Update the “find_those_words” option in the details section.

aresxs91 commented:
The link (https://core.ac.uk/docs/#!/all/advanced_search) is actually not public so I've included a description of the information needed in the data frame for advanced search queries in the details section.

Caught red-handed on the line widths!
Agreed with the is.data.frame check.

aristotelisxs committed:
roxygen2 still persistent in not generating results after instructions followed here: http://stackoverflow.com/questions/33529903/roxygen2-version-5-0-incorrectly-creates-documentation-when-occurs-inside-f
Omitting them for now.

sckott commented:
thanks @aresxs91 -- all done with changes?

aresxs91 commented:
Yes, all done, let me know if you spot anything I have missed. Thanks, @sckott

sckott commented on R/core_advanced_search.R:
```diff
@@ -0,0 +1,74 @@
+               list(page = page, pageSize = limit), ...)
+    }
+  } else {
+    print("query must be a data.frame (see details)")
```
I wouldn't use `print()` here. Use `stop()` instead, can can just move that up above the block of code like 

```r
stop("query must be a data.frame (see details)")

parsed_advanced_query <- apply(X = query, MARGIN = 1, FUN = 
                                 parse_advanced_search_query)

if(length(parsed_advanced_query) > 1){
  queries <- create_batch_query_list(parsed_advanced_query, page, limit)
  args <- NULL
  
  core_POST(path = "search", key, args, queries, ...)
} else {
  core_GET(path = file.path("search", parsed_advanced_query), key,
           list(page = page, pageSize = limit), ...)
}
```

aristotelisxs replied:
True, missed that while debugging.

sckott commented on R/core_advanced_search.R:
```diff
@@ -0,0 +1,74 @@
+#' optional
+#' @details `query` should include columns with the following information 
+#' (at least one is required):
+#' 1. `all_of_the_words`: string, with space separated terms that should all exist
```
there's still 5 lines that have widths > 80 chars. including this one

sckott commented on R/core_advanced_search.R:
```diff
@@ -0,0 +1,74 @@
+#' indicated
+#' `core_advanced_search` does the HTTP request and parses, while
+#' `core_advanced_search_` just does the HTTP request, gives back JSON as a 
+#' character string
```
add a `@return` tag and describe what is returned

aristotelisxs replied:
I haven't seen any other of the functions have a @return tag so I will add it in for them too.

Just to mention that the advanced search return the same data type as the rest, a data frame with:
1) "status" as character, 
2) "totalHits" as integer and
3) "data" as list.

aristotelisxs replied:
Adding to my previous comment, I missed that the references tag would actually cover for what the other endpoints return, whereas the advanced search does not have one. I see why you wanted a return tag here, will add it only for this endpoint.

sckott replied:
Thanks for adding `@return` (sometimes I forget them too)

that sounds good what you described for return

sckott commented on R/zzz.R:
```diff
@@ -1,5 +1,192 @@
 core_base <- function() "https://core.ac.uk"
 
+# Advanced search related functions
```
- move all these helper functions into a new file with file name indicating what they are, please use `.R` and not `.r`
- be consistent with fxn names, ideally `snake_case`, right now there's combo of camelCase, snake_case and dot.case

aristotelisxs replied:
Agreed, I will move all advanced search related functions.
Will rename fxn_names accordingly

sckott commented on R/core_advanced_search.R:
```diff
@@ -0,0 +1,74 @@
+#' 
+#' res <- core_advanced_search(query)
+#' head(res$data)
+#' res$data[[1]]$id
```
you need a closing bracket `}` on a new line after this last examples line

sckott commented on R/zzz.R:
```diff
@@ -1,5 +1,192 @@
+    if (na.rm == T) {
+      paste.na <- function(x, sep) {
+        x <- gsub("^\\s+|\\s+$", "", x)
+        ret <- paste(na.omit(x), collapse = sep)
```
namespace all calls to non-base R pkgs like `stats`, `tools`, etc. 

here, namespace like `stats::na.omit` instead of `na.omit`

sckott reviewed (changes requested):
looking good, a few more things to change. 

also, can you add some tests in a new test file for the adv. function.

aresxs91 commented:
Thanks for you comments again @sckott, as for the advanced search query tests, I guess an approach would be to add a few queries using a random combination of available filters. Would that be enough?

aristotelisxs committed:
Several changes, including:
1) Moving advanced search related functions to their own file, namely advanced_search_helper_functions
2) Adding a description of what’s returned by the core_advanced_search endpoint
3) Advanced search test file
4) Changes on how the year filter option operates to first check if we are dealing with an NA before passing it through the further validity checks -> The “further validity checks” need to be minimised as they are some redundancies now with the introduction of this that were specific only to Angular2 previously (from where they were translated)
5) Line width and function names cleaning up

aristotelisxs committed:
Absent year filters should be identified through the data frame’s “names” property instead.

sckott commented:
Thanks for adding some tests, we can adjust those as we see fit later. 

Did you do all changes?

aresxs91 commented:
I see, great then.

I want to have a closer look at some of the validity checks, not all of them are needed so I'll need some more time on that. Will let you know, thanks!

aresxs91 commented:
I see, great then.

I want to have a closer look at some of the validity checks, not all of them are needed so I'll need some more time on that. Will let you know, thanks!

sckott commented:
okay

aresxs91 commented:
Hi @sckott, after a quick review, there shouldn't be any immediate changes I'll need to do on the filter values validity checks. There may be edge cases not captured but I think for the most part, they should be ready to go. If there's nothing else to add, let's get this into production!

aristotelisxs committed:
Small update on function names.

sckott commented:
okay, looks good
